risc0/circuit/rv32im/src/execute/bigint/analyze.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2025 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::Result;
+use byteorder::{LittleEndian, ReadBytesExt as _};
+
+use std::collections::HashMap;
+
+use crate::execute::{
+    bibc::{OpCode, Program},
+    pager::{NODE_CYCLES, PAGE_CYCLES},
+    WORD_SIZE,
+};
+
+const PAGE_SIZE: u32 = 1024;
+
+/// What is the maximum number of pages an IO of the given size can span?
+fn max_pages_for_io(size: u32) -> u32 {
+    if size == 0 {
+        0
+    } else if size == 1 {
+        1
+    } else {
+        let p = size / PAGE_SIZE;
+        match size % PAGE_SIZE {
+            0 | 1 => 2 + (p - 1),
+            _ => 2 + p,
+        }
+    }
+}
+
+#[test]
+fn max_pages_for_io_test() {
+    assert_eq!(max_pages_for_io(0), 0);
+    assert_eq!(max_pages_for_io(1), 1);
+    assert_eq!(max_pages_for_io(2), 2);
+
+    assert_eq!(max_pages_for_io(1024), 2);
+    assert_eq!(max_pages_for_io(1025), 2);
+    assert_eq!(max_pages_for_io(1026), 3);
+    assert_eq!(max_pages_for_io(1027), 3);
+
+    assert_eq!(max_pages_for_io(1024 * 2), 3);
+    assert_eq!(max_pages_for_io(1024 * 2 + 1), 3);
+    assert_eq!(max_pages_for_io(1024 * 2 + 2), 4);
+}
+
+struct IoCounter {
+    cycles: u32,
+    loaded_arenas: HashMap<u32, u32>,
+    dirty_arenas: HashMap<u32, u32>,
+}
+
+impl IoCounter {
+    fn new() -> Self {
+        Self {
+            cycles: 0,
+            loaded_arenas: HashMap::new(),
+            dirty_arenas: HashMap::new(),
+        }
+    }
+
+    fn track_load_arena(&mut self, arena: u32, offset: u32, count: u32) {
+        let mut page_in = false;
+
+        let num_pages = self.loaded_arenas.entry(arena).or_insert(0);
+
+        while *num_pages < max_pages_for_io(offset + count) {
+            // load could go all the way up the tree
+            self.cycles += PAGE_CYCLES;
+            self.cycles += 21 * NODE_CYCLES;
+            page_in = true;
+            *num_pages += 1;
+        }
+
+        println!("load_arena({arena}, {offset}, {count}), page_in = {page_in}");
+    }
+
+    fn track_store_arena(&mut self, arena: u32, offset: u32, count: u32) {
+        self.cycles += PAGE_CYCLES;
+        self.cycles += 22 * NODE_CYCLES * 2;
```

Comment:
I'm not sure about this line. Could you describe what's going on here? Why 22 instead of 21? Is store arena in kernel memory?

Reply:
This being `22` instead of `21` was just a mistake I think. The `21` think is the height of the tree - 1, and I originally had `22` for all of these but missed fixing this one.
It isn't in kernel memory.